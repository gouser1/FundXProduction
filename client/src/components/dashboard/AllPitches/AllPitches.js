import React from "react";
import {
  IconButton,
  Avatar,
  Card,
  CardHeader,
  CardMedia,
  CardContent,
  Box,
  CardActions,
  Badge,
  Grid,
  Typography,
  Paper,
  Container,
  Button as NormalButton,
} from "@material-ui/core";
import axios from "axios";
import { useEffect, useState } from "react";
import { useHistory, Link } from "react-router-dom";
import Favorite from "@material-ui/icons/Favorite";
import Location from "@material-ui/icons/LocationOn";
import industries from "../../../data/industries.json"; // Form data fields
import countries from "../../../data/countriesFilter.json";
import cardImage from "../../../images/dashboard/placeholder.png";
import userIcon from "../../../images/dashboard/usericon.png";
import { Formik, Form } from "formik";
import Select from "../../FormsUI/Select/index"; // Form Select
import Button from "../../FormsUI/Button/index"; // Form Submit button
import useStyles from "./AllPitchesStyle"; // Component Styles

const AllPitches = () => {
  const [listOfPitches, setListOfPitches] = useState([]);
  const [favouritedPitches, setFavouritedPitches] = useState([]);
  let history = useHistory();
  const classes = useStyles();

  useEffect(() => {
    getListOfPitches();
  }, []);

  // Get all pitches
  const getListOfPitches = () => {
    axios({
      method: "get",
      url: "http://localhost:3001/pitches",
      headers: { accessToken: localStorage.getItem("accessToken") },
    }).then((response) => {
      setListOfPitches(response.data.pitchList);
    });
  };

  // Get filtered list of pitches
  const getFilteredListOfPitches = (data) => {
    let country = data.country === "All" ? "" : data.country;
    let industry = data.industry;

    axios({
      method: "get",
      // encodeURIComponent used as a querystring parameter
      url: `http://localhost:3001/pitches/?country=${country}&industry=${encodeURIComponent(
        industry
      )}`,
      headers: { accessToken: localStorage.getItem("accessToken") },
    }).then((response) => {
      // set filtered list response to pitch list
      setListOfPitches(response.data.pitchList);
    });
  };

  // Sort by capital needed in descending order
  const sortByLargestCapitalNeeded = () => {
    let newListOfPitches = listOfPitches.sort(
      (a, b) => b.capitalNeeded - a.capitalNeeded
    );
    setListOfPitches([...newListOfPitches]);
  };

  // Sort  by capital needed in ascending order
  const sortBySmallestCapitalNeeded = () => {
    let newListOfPitches = listOfPitches.sort(
      (a, b) => a.capitalNeeded - b.capitalNeeded
    );
    setListOfPitches([...newListOfPitches]);
  };

  // Favourite Pitch
  const favouritePitch = (pitchId) => {
    axios({
      method: "post",
      url: "http://localhost:3001/favourite",
      data: {
        PitchId: pitchId,
      },
      headers: { accessToken: localStorage.getItem("accessToken") },
    }).then((response) => {
      setListOfPitches(
        listOfPitches.map((pitch) => {
          if (pitch.id === pitchId) {
            if (response.data.favourited) {
              // Destructure pitch and modify favourites field and adding 0 so the length is modified to favourite the post
              return { ...pitch, Favourites: [...pitch.Favourites, 0] };
            } else {
              // Remove last element from array to update state of favourited post to remove favourite
              const FavouritesArray = pitch.Favourites;
              FavouritesArray.pop();
              return { ...pitch, Favourites: FavouritesArray };
            }
          } else {
            return pitch;
          }
        })
      );
      if (favouritedPitches.includes(pitchId)) {
        setFavouritedPitches(
          favouritedPitches.filter((id) => {
            return id !== pitchId;
          })
        );
      } else {
        setFavouritedPitches([...favouritedPitches, pitchId]);
      }
    });
  };

  // Formik Initial Form State
  const INITIAL_FORM_STATE = {
    country: "",
    industry: "",
  };

  return (
    <div className={classes.root}>
      <Grid
        container
        justifyContent="center"
        style={{ minHeight: "100vh", maxWidth: "100%" }}
      >
        <Grid item xs={12} style={{ paddingTop: "2%" }}>
          <Container maxWidth="lg">
            <Formik
              initialValues={INITIAL_FORM_STATE}
              onSubmit={(data) => {
                getFilteredListOfPitches(data);
              }}
            >
              <Form>
                <Typography
                  className={classes.h1}
                  style={{ paddingBottom: "5%" }}
                >
                  All Pitches
                </Typography>
                <Typography
                  className={classes.p1}
                  style={{ paddingBottom: "1%" }}
                >
                  Industries
                </Typography>
                <Select name="industry" options={industries}></Select>
                <Typography
                  className={classes.p1}
                  style={{ paddingBottom: "1%", paddingTop: "1%" }}
                >
                  Countries
                </Typography>
                <Select
                  name="country"
                  options={countries}
                  style={{ paddingBottom: "2%" }}
                ></Select>
                <Button className={classes.button}>
                  <Typography className={classes.p1} style={{ color: "white" }}>
                    Go!
                  </Typography>
                </Button>
                <NormalButton
                  className={classes.button}
                  style={{ backgroundColor: "#f44336", marginLeft: "1%" }}
                  onClick={() => {
                    getListOfPitches();
                  }}
                >
                  <Typography style={{ color: "white" }}>Reset</Typography>
                </NormalButton>
              </Form>
            </Formik>
            <Box
              style={{ marginTop: "2%", marginBottom: "5%", display: "flex" }}
            >
              <NormalButton
                style={{ marginRight: "2%" }}
                size="small"
                color="primary"
                variant="outlined"
                className={classes.buttonOutline}
                onClick={() => {
                  sortByLargestCapitalNeeded();
                }}
              >
                Largest Capital Needed
              </NormalButton>
              <NormalButton
                size="small"
                color="primary"
                variant="outlined"
                className={classes.buttonOutline}
                onClick={() => {
                  sortBySmallestCapitalNeeded();
                }}
              >
                Smallest Capital Needed
              </NormalButton>
            </Box>
            {/* List of all pitches displayed here */}
            <Grid container spacing={3}>
              {listOfPitches.map((value, key) => {
                return (
                  <Grid item xs={12} md={4} lg={4}>
                    {/* Pitch Card Start */}
                    <Paper className={classes.paper}>
                      <Card className={classes.card}>
                        <CardContent>
                          <CardHeader
                            avatar={
                              <Link to={`profile/${value.UserId}`}>
                                <Avatar>
                                  <img
                                    src={userIcon}
                                    alt=""
                                    width="50px"
                                    height="35px"
                                  />
                                </Avatar>
                              </Link>
                            }
                            titleTypographyProps={{ variant: "h6" }}
                            title={value.pitchTitle}
                            subheader={
                              <Badge className={classes.badge}>
                                {value.location}, {value.country}
                                <Location />{" "}
                              </Badge>
                            }
                          />
                          <CardMedia
                            className={classes.media}
                            image={cardImage}
                          />
                          <Typography
                            className={classes.p1}
                            style={{ marginTop: "3%" }}
                          >
                            {value.pitchInfo.length > 100
                              ? `${value.pitchInfo.substring(0, 100)}...`
                              : value.pitchInfo}
                          </Typography>
                          <div
                            style={{
                              display: "flex",
                              justifyContent: "space-between",
                              paddingTop: "0.5em",
                            }}
                          >
                            <Typography
                              className={classes.p1}
                              style={{ marginTop: "3%" }}
                            >
                              Amount raised: £{value.capitalRaised}
                            </Typography>

                            <Typography
                              className={classes.p1}
                              style={{ marginTop: "3%" }}
                            >
                              Amount needed: £{value.capitalNeeded}
                            </Typography>
                          </div>
                          <Typography
                            className={classes.p1}
                            style={{ marginTop: "3%", fontSize: "0.9em" }}
                          >
                            Industry: {value.industry}
                            {""}
                            {value.industry && value.industry2 ? "," : ""}{" "}
                            {value.industry2}
                          </Typography>
                        </CardContent>

                        <CardActions
                          style={{
                            display: "flex",
                            justifyContent: "space-between",
                          }}
                        >
                          <NormalButton
                            size="small"
                            className={classes.button}
                            style={{ marginTop: "3%" }}
                            onClick={() => {
                              history.push(`/dashboard/pitch/${value.id}`);
                            }}
                          >
                            <Typography style={{ color: "white" }}>
                              More Info
                            </Typography>
                          </NormalButton>

                          <IconButton
                            aria-label="add to favorites"
                            onClick={() => {
                              favouritePitch(value.id);
                            }}
                          >
                            <Favorite
                              style={{
                                color: favouritedPitches.includes(value.id)
                                  ? "#f44336"
                                  : "#f44336",
                              }}
                              className={
                                favouritedPitches.includes(value.id)
                                  ? "unfavouriteButton"
                                  : "FavouriteButton"
                              }
                            />

                            <Typography>{value.Favourites.length}</Typography>
                          </IconButton>
                        </CardActions>
                      </Card>
                    </Paper>
                    {/* Pitch Card End */}
                  </Grid>
                );
              })}
            </Grid>
          </Container>
        </Grid>
      </Grid>
    </div>
  );
};

export default AllPitches;
