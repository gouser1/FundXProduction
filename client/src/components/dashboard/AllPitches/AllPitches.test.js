import React from "react";
import renderer from "react-test-renderer";

import AllPitches from "./AllPitches";

it("renders correctly", () => {
  const tree = renderer.create(<AllPitches />).toJSON();
  expect(tree).toMatchSnapshot();
});
