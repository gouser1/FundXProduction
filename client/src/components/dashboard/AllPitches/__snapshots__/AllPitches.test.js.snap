// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="makeStyles-root-1"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-justify-content-xs-center"
    style={
      Object {
        "maxWidth": "100%",
        "minHeight": "100vh",
      }
    }
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      style={
        Object {
          "paddingTop": "2%",
        }
      }
    >
      <div
        className="MuiContainer-root MuiContainer-maxWidthLg"
      >
        <form
          action="#"
          onReset={[Function]}
          onSubmit={[Function]}
        >
          <p
            className="MuiTypography-root makeStyles-h1-6 MuiTypography-body1"
            style={
              Object {
                "paddingBottom": "5%",
              }
            }
          >
            All Pitches
          </p>
          <p
            className="MuiTypography-root makeStyles-p1-5 MuiTypography-body1"
            style={
              Object {
                "paddingBottom": "1%",
              }
            }
          >
            Industries
          </p>
          <div
            className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              onClick={[Function]}
            >
              <div
                aria-haspopup="listbox"
                aria-labelledby="mui-component-select-industry"
                className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
                id="mui-component-select-industry"
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                role="button"
                tabIndex={0}
              >
                <span
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "&#8203;",
                    }
                  }
                />
              </div>
              <input
                aria-hidden={true}
                autoFocus={false}
                className="MuiSelect-nativeInput"
                name="industry"
                onAnimationStart={[Function]}
                onChange={[Function]}
                required={false}
                tabIndex={-1}
                value=""
              />
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <fieldset
                aria-hidden={true}
                className="PrivateNotchedOutline-root-12 MuiOutlinedInput-notchedOutline"
                style={
                  Object {
                    "paddingLeft": 8,
                  }
                }
              >
                <legend
                  className="PrivateNotchedOutline-legend-13"
                  style={
                    Object {
                      "width": 0.01,
                    }
                  }
                >
                  <span
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "&#8203;",
                      }
                    }
                  />
                </legend>
              </fieldset>
            </div>
          </div>
          <p
            className="MuiTypography-root makeStyles-p1-5 MuiTypography-body1"
            style={
              Object {
                "paddingBottom": "1%",
                "paddingTop": "1%",
              }
            }
          >
            Countries
          </p>
          <div
            className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            style={
              Object {
                "paddingBottom": "2%",
              }
            }
          >
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              onClick={[Function]}
            >
              <div
                aria-haspopup="listbox"
                aria-labelledby="mui-component-select-country"
                className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
                id="mui-component-select-country"
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                role="button"
                tabIndex={0}
              >
                <span
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "&#8203;",
                    }
                  }
                />
              </div>
              <input
                aria-hidden={true}
                autoFocus={false}
                className="MuiSelect-nativeInput"
                name="country"
                onAnimationStart={[Function]}
                onChange={[Function]}
                required={false}
                tabIndex={-1}
                value=""
              />
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <fieldset
                aria-hidden={true}
                className="PrivateNotchedOutline-root-12 MuiOutlinedInput-notchedOutline"
                style={
                  Object {
                    "paddingLeft": 8,
                  }
                }
              >
                <legend
                  className="PrivateNotchedOutline-legend-13"
                  style={
                    Object {
                      "width": 0.01,
                    }
                  }
                >
                  <span
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "&#8203;",
                      }
                    }
                  />
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
            disabled={false}
            fullwidth={true}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              <p
                className="MuiTypography-root makeStyles-p1-5 MuiTypography-body1"
                style={
                  Object {
                    "color": "white",
                  }
                }
              >
                Go!
              </p>
            </span>
          </button>
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-button-3"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "backgroundColor": "#f44336",
                "marginLeft": "1%",
              }
            }
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              <p
                className="MuiTypography-root MuiTypography-body1"
                style={
                  Object {
                    "color": "white",
                  }
                }
              >
                Reset
              </p>
            </span>
          </button>
        </form>
        <div
          className="MuiBox-root MuiBox-root-16"
          style={
            Object {
              "display": "flex",
              "marginBottom": "5%",
              "marginTop": "2%",
            }
          }
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-buttonOutline-4 MuiButton-outlinedPrimary MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "marginRight": "2%",
              }
            }
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              Largest Capital Needed
            </span>
          </button>
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-buttonOutline-4 MuiButton-outlinedPrimary MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              Smallest Capital Needed
            </span>
          </button>
        </div>
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
        />
      </div>
    </div>
  </div>
</div>
`;
